fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SingleSidePredicate.java

```diff
@@ -0,0 +1,112 @@
+    public Rule build() {
+        return logicalJoin()
+                .when(join -> join.getHashJoinConjuncts().stream().anyMatch(expr ->
+                        expr.children().stream().anyMatch(expr1 -> !(expr1 instanceof Slot))))
+                .then(join -> {
+                    List<List<Expression>> exprsOfJoinRelation =
+                            Lists.newArrayList(Lists.newArrayList(), Lists.newArrayList());
+                    Map<Expression, Alias> exprMap = Maps.newHashMap();
+                    Plan left = join.left();
+                    join.getHashJoinConjuncts().forEach(conjunct -> {
+                        Preconditions.checkArgument(conjunct instanceof EqualTo);
+                        Expression[] exprs = conjunct.children().toArray(new Expression[0]);
+                        // sometimes: t1 join t2 on t2.a + 1 = t1.a + 2, so check the situation, but actually it
+                        // doesn't swap the two sides.
+                        int tag = checkIfSwap(exprs[0], left);
+                        exprsOfJoinRelation.get(0).add(exprs[tag]);
+                        exprsOfJoinRelation.get(1).add(exprs[tag ^ 1]);
+                        Arrays.stream(exprs).sequential().forEach(expr ->
+                                exprMap.put(expr, new Alias(expr, "expr_" + expr.hashCode())));
+                    });
+                    Iterator<List<Expression>> iter = exprsOfJoinRelation.iterator();
+                    return join.withhashJoinConjunctsAndChildren(
+                            join.getHashJoinConjuncts()
+                                    .stream().map(equalTo -> equalTo.withChildren(equalTo.children()
+                                            .stream().map(expr -> exprMap.get(expr).toSlot())
+                                            .collect(Collectors.toList())))
+                                    .collect(Collectors.toList()),
+                            join.children().stream().map(
+                                    plan -> new LogicalProject<>(new ImmutableList.Builder<NamedExpression>()
+                                            .addAll(iter.next().stream().map(expr -> exprMap.get(expr))
+                                                    .collect(Collectors.toList()))
+                                            .addAll(getOutput(plan, join)).build(), plan))
+                                    .collect(Collectors.toList()));
+                }).toRule(RuleType.PUSH_DOWN_NOT_SLOT_REFERENCE_EXPRESSION);
+    }
+
+    int checkIfSwap(Expression left, Plan joinLeft) {
+        Set<Expression> joinOut = ImmutableSet.copyOf(joinLeft.getOutput());
+        return left.anyMatch(expr -> (expr instanceof Slot) && joinOut.contains(expr)) ? 0 : 1;
+    }
+
+    private List<Slot> getOutput(Plan plan, LogicalJoin join) {
+        Set<Slot> set = Sets.newHashSet(plan.getOutputSet());
+        set.retainAll(join.getOutputSet());
+        return Arrays.asList(set.toArray(new Slot[0]));
```

Comment:
if u wanna turn set into list. just use Lists.newArrayList(set);
BTW, `set` is not a good name